---
title: 'SISBD: GEO lab'
author: "Raphael Gottardo"
date: "\ `r Sys.Date()`"
output:
  ioslides_presentation:
    fig_caption: yes
    fig_retina: 1
    keep_md: yes
    smaller: yes
---

## Setting up some options

Let's first turn on the cache for increased performance and improved styling
```{r, cache=FALSE}
# Set some global knitr options
library("knitr")
opts_chunk$set(cache = FALSE, messages = FALSE)
```

Load the `pander` package so we can make nicer table listings with `pandoc.table`.

```{r}
suppressMessages(library(pander))
```

## Preparation

We will need to query the GEOmetadb database. Today we will explore this 
database and practice various ways to query it.

## Load the `GEOmetadb` package

First we load the `GEOmetadb` library.

```{r}
suppressMessages(library(GEOmetadb))
```

Let's also view the available objects.

```{r}
ls("package:GEOmetadb")
```

## Download the GEO database

We should have already downloaded this database when viewing the lecture slides.

```{r}
## This will download the entire database, so can be slow
if (!file.exists("GEOmetadb.sqlite"))
{
  # Download database only if it's not done already
  getSQLiteFile()
}
```

## List tables with `SQL`

In `SQL`, you can query the database structure with ordinary `SQL` commands.

```{r}
geo_con <- dbConnect(SQLite(), 'GEOmetadb.sqlite')
dbGetQuery(geo_con, "SELECT name FROM sqlite_master WHERE type='table';")
```

## List `gse` fields with `SQL`

The `PRAGMA` command is a standard `SQLite` command.

```{r}
dbGetQuery(geo_con, "PRAGMA table_info(gse);")
```

## List tables with `dbListTables`

Instead of using `SQL` commands, we can list tables and fields with functions
from the `GEOmetadb` package.

```{r}
geo_con <- dbConnect(SQLite(), 'GEOmetadb.sqlite')
dbListTables(geo_con)
```

```{r}
dbListFields(geo_con, 'gse')
```

## Explore `gse`

```{r}
columnDescriptions()[1:5,]
```

## Load library `data.table`

This will provide us with some practice querying with data.table.

```{r}
suppressMessages(library(data.table))
```

## Explore `gse` with `data.table`

```{r}
cd <- as.data.table(columnDescriptions())
cd[TableName == "gse", FieldName]
```

## List `gse` columns with `pandoc.table`

```{r}
gsefields <- as.data.frame(
    cd[TableName == "gse" & 
           FieldName %in% c("gse","title","pubmed_id","summary","contact")])
pandoc.table(gsefields, style = "grid")
```

## Explore `gpl`

```{r}
cd[TableName == "gpl", FieldName]
```

## Explore columns in `gpl`

```{r}
gplfields <- as.data.frame(
    cd[TableName == "gpl" & 
           FieldName %in% c("gpl", "organism", "manufacturer")])
pandoc.table(gplfields, style = "grid")
```

## Explore `gse_gpl`

```{r}
cd[TableName == "gse_gpl", FieldName]
```

## Explore columns in `gse_gpl`

Why are there only two fields in this table? What is this table for?

```{r}
gse_gplfields <- as.data.frame(cd[TableName == "gse_gpl"])
pandoc.table(gse_gplfields, style = "grid")
```

## List "title" fields with `pandoc.table`

Why do many tables include a "title" field? Are the titles the same?

```{r}
gsefields <- as.data.frame(
    cd[FieldName == "title"])
pandoc.table(gsefields, style = "grid")
```

## List "contact" field structure

Let's look at some records in `gse`. What does a "contact" look like?

```{r}
query <- "SELECT contact FROM gse LIMIT 1;"
res <- dbGetQuery(geo_con, query)
strsplit(res$contact, '\t')
```

## Find manufacturer data

Query the manufacturers with a `SQL` command, listed with `data.table`...

```{r, tidy=FALSE}
manu <- data.table(dbGetQuery(geo_con, 
    "SELECT DISTINCT manufacturer FROM gpl ORDER BY manufacturer ASC;"))
manu[,list(length(manufacturer)), by = manufacturer]
```

## Our `SQL` command

We just wanted a list of manufacturers so the `SQL` query is:

```
SELECT DISTINCT manufacturer FROM gpl 
ORDER BY manufacturer ASC;
```

However, since we also grouped `by=manufacturer` in our `data.table`, we could 
have simply used the `SQL` query:

```
SELECT manufacturer FROM gpl;
```

Let's try that...

## Find manufacturer data

Query the manufacturers with a simpler `SQL` command ... grouping with `by` and 
ordering with `setkey` in `data.table`...

```{r, tidy=FALSE}
manu <- data.table(dbGetQuery(geo_con, 
            "SELECT manufacturer FROM gpl;"))
setkey(manu, manufacturer)
manu[, list(length(manufacturer)), by = manufacturer]
```


## Finding data with a `join`

To get supplementary file names ending with `CEL.gz` (case-insensitive) from 
only manufacturer Affymetrix, we need to `join` the `gsm` and `gpl` tables. 

```
SELECT 
        gpl.bioc_package, 
        gsm.title, 
        gsm.series_id, 
        gsm.gpl, 
        gsm.supplementary_file 
    FROM gsm 
    JOIN gpl ON gsm.gpl=gpl.gpl 
    WHERE gpl.manufacturer='Affymetrix' 
        AND gsm.supplementary_file like '%CEL.gz';
```

## Now let's run that query

```{r, tidy=FALSE}
query <- "SELECT 
            gpl.bioc_package, 
            gsm.title, 
            gsm.series_id, 
            gsm.gpl, 
            gsm.supplementary_file 
        FROM gsm 
        JOIN gpl ON gsm.gpl=gpl.gpl 
        WHERE gpl.manufacturer='Affymetrix' 
            AND gsm.supplementary_file like '%CEL.gz';"
res <- dbGetQuery(geo_con, query)
head(res, 3)
```

## Why did we need a `join`?

The 
[GEOmetadb database](http://gbnci.abcc.ncifcrf.gov/geo/geo_help.php), 
is a [relational database](http://en.wikipedia.org/wiki/Relational_database). 

There are several tables which can be linked on common fields. 

Since each table 
contains data for only one type of record, tables must be linked to search for 
fields pertaining to the various types of records. 

We join on the common fields, 
called [keys](http://en.wikipedia.org/wiki/Relational_database#Primary_key).

## Table Relationships of `GEOmetadb`

![Table Relationships](http://gbnci.abcc.ncifcrf.gov/geo/images/GEOmetadb_diagram.png)

Source: [Help: GEOmetadb Application, Meltzerlab/GB/CCR/NCI/NIH &copy;2008](http://gbnci.abcc.ncifcrf.gov/geo/geo_help.php)

## Keys of `GEOmetadb`

```
+------------+-------+------------------------------------------------+
| Table      | Key   | Links to Table.Key                             |
+============+=======+================================================+
| gse        | gse   | gse_gpl.gse, gse_gsm.gse, gds.gse, sMatrix.gse |
+------------+-------+------------------------------------------------+
| gpl        | gpl   | gds.gpl, gse_gpl.gpl, sMatrix.gpl, gsm.gpl     |
+------------+-------+------------------------------------------------+
| gsm        | gsm   | gse_gsm.gsm                                    |
| gsm        | gpl   | gds.gpl, gse_gpl.gpl, sMatrix.gpl, gpl.gpl     |
+------------+-------+------------------------------------------------+
| gds        | gds   | gds_subset.gds                                 |
+------------+-------+------------------------------------------------+
| gds_subset | gds   | gds.gds                                        |
+------------+-------+------------------------------------------------+
| sMatrix    | gse   | gse_gpl.gse, gse_gsm.gse, gds.gse, gse.gse     |
| sMatrix    | gpl   | gds.gpl, gse_gpl.gpl, gpl.gpl, gsm.gpl         |
+------------+-------+------------------------------------------------+
| gse_gpl    | gse   | gse_gpl.gse, gse_gsm.gse, gds.gse, sMatrix.gse |
| gse_gpl    | gpl   | gds.gpl, gse_gpl.gpl, gpl.gpl, sMatrix.gpl     |
+------------+-------+------------------------------------------------+
| gse_gsm    | gse   | gse_gpl.gse, gse.gse, gds.gse, sMatrix.gse     |
| gse_gsm    | gsm   | gsm.gsm                                        |
+------------+-------+------------------------------------------------+
```

Source: [Help: GEOmetadb Application, Meltzerlab/GB/CCR/NCI/NIH &copy;2008](http://gbnci.abcc.ncifcrf.gov/geo/geo_help.php)

## A three-table `join`

To get raw data, we need to `join` three tables with two `join` clauses. The first
`join` is a subquery in the `from` clause, using `gse_gsm` to find `gsm` records
corresponding to `gse` records. We then `join` this with `gsm` for those records. 
This approach works well when you only have a few queries to make or you have 
limited memory (RAM) available.

```{r, tidy=FALSE}
query <- "SELECT gsm.gsm, gsm.supplementary_file 
        FROM (gse JOIN gse_gsm ON gse.gse=gse_gsm.gse) j 
        JOIN gsm ON j.gsm=gsm.gsm 
        WHERE gse.pubmed_id='21743478' 
        LIMIT 2;"
res <- as.data.table(dbGetQuery(geo_con, query))
res[, strsplit(gsm.supplementary_file, ';\t'), by = gsm.gsm]
```

## Joins in `data.table`

We can repeat the same operation using `data.table`, once we have converted the 
GEO tables to `data.table`s and set their keys. 

```{r, tidy=FALSE}
gseDT <- data.table(dbGetQuery(geo_con, "SELECT * from gse;"), key = "gse")
gsmDT <- data.table(dbGetQuery(geo_con, "SELECT * from gsm;"), key = "gsm")
gse_gsmDT <- data.table(dbGetQuery(geo_con, "SELECT * from gse_gsm;"), 
    key = c("gse", "gsm"))
gsmDT[gse_gsmDT[gseDT[pubmed_id == 21743478, gse], gsm, nomatch = 0], 
      nomatch = 0][1:2, list(gsm, supplementary_file)][,strsplit(supplementary_file, ';\t'), by = gsm]
```

## All in one line?

Can we do it all in one line of code? Yes, but it's ugly and hard to follow, 
even with line-wrap. Plus, additional queries will have to reload the data from 
the database. Yuk! (Don't do it this way.)

```{r, tidy=FALSE}
data.table(dbGetQuery(geo_con, 
    "SELECT * from gsm;"), key = "gsm")[data.table(dbGetQuery(geo_con, 
    "SELECT * from gse_gsm;"), key = c("gse", "gsm"))[data.table(dbGetQuery(geo_con, 
    "SELECT * from gse;"), key = "gse")[pubmed_id == 21743478, gse], gsm, 
    nomatch = 0], nomatch = 0][1:2, list(gsm, supplementary_file)][,
    strsplit(supplementary_file, ';\t'), by = gsm]
```

## Joining with `merge`

Some people like to use the familiar `merge`. There is a version of `merge`
built into `data.table` for improved performance. We will use the three DTs we 
made previously. To remove duplicates, we use `unique`. (Why are there duplicates?)

```{r, tidy=FALSE}
unique(merge(gsmDT[,list(gsm,supplementary_file)], 
      merge(gseDT[pubmed_id == 21743478, list(gse)], 
            gse_gsmDT)[,list(gsm)])[1:4, list(gsm, supplementary_file)])[,
                    strsplit(supplementary_file, ';\t'), by = gsm]
```

## Joining with `merge` and `magrittr`

We can also use `%>%` from `magrittr` to improve readability, again using the 
three DTs we made previously. Here we will use two "lines" of code.

```{r, tidy=FALSE}
library(magrittr)
mergedDT <- unique(gseDT[pubmed_id == 21743478, list(gse)] %>% 
                merge(y = gse_gsmDT, by = c("gse")) %>% 
                merge(y = gsmDT[,list(gsm,supplementary_file)], by = c("gsm")))
mergedDT[1:2, list(gsm, gse, supplementary_file)][,
                strsplit(supplementary_file, ';\t'), by = gsm]
```

## Only get what you need

It makes sense to only `select` the data we need from the SQL database. Why pull 
in extra data, only to ignore it? We will still use `data.table` for the `join`, 
though, in keeping with the spirit of the assignment.

```{r, tidy=FALSE}
gseDT <- data.table(dbGetQuery(geo_con, 
    "SELECT gse from gse WHERE pubmed_id = '21743478';"), key = "gse")
gsmDT <- data.table(dbGetQuery(geo_con, 
    "SELECT gsm, supplementary_file from gsm;"), key = "gsm")
gse_gsmDT <- data.table(dbGetQuery(geo_con, 
    "SELECT * from gse_gsm;"), key = c("gse", "gsm"))
gsmDT[gse_gsmDT[gseDT, gsm, nomatch = 0], nomatch = 0][1:2, 
    list(gsm, supplementary_file)][,strsplit(supplementary_file, ';\t'), by = gsm]
```

## Cleanup

```{r}
dbDisconnect(geo_con)
```

## Time for you to work!

1. Use the `GEOmetabd` package to find all HCV gene expression data using the Illumina platform submitted by an investigator at Yale. This should be done with a single query, showing the title, the GSE accession number, the GPL accession number and the manufacturer and the description of the platform used.

2. Reproduce your above query using the data.table package. Again, try to use a single line of code. [Hint: You first need to convert all db tables to data.table tables].


# Solution

## Let's create a new connection

```{r}
con <- dbConnect(SQLite(),'GEOmetadb.sqlite')
```

## Using SQL

```{r}
res <- dbGetQuery(con, "SELECT gse.title, gse.gse, gpl.gpl, gpl.manufacturer, gpl.title
                  FROM (gse JOIN gse_gpl ON gse.gse=gse_gpl.gse) j
                  JOIN gpl on j.gpl=gpl.gpl 
                  WHERE gpl.manufacturer like '%Illumina%'
                  AND gse.contact like '%Yale%'
                  AND (gse.title like '%epatit%' OR gse.title like '%HCV%'); ")
```

## Using data.table

```{r}
gse <- data.table(dbGetQuery(con, "SELECT gse.title, gse.gse, gse.contact FROM gse"), key = "gse")
gpl <- data.table(dbGetQuery(con, "SELECT gpl.gpl, gpl.manufacturer, gpl.description FROM gpl"), key = "gpl")
gse_gpl <- data.table(dbGetQuery(con, "SELECT gse_gpl.gse, gse_gpl.gpl FROM gse_gpl"), key = "gse")

res2 <- merge(gse[gse_gpl, nomatch = 0], gpl, by = "gpl")[title %like% "HCV" 
                                                          & contact %like% "Yale" & manufacturer %like% "Illumina"]
```

## Now what?

We have identifed the data we want, let's download them

```{r}
gds <- getGEO("GSE40812", destdir = "data/geo/")[[1]]
pData(gds)[1:2,]
```

## Sanitize our data

```{r}
sanitize_pdata <- function(pd) {
keepCols <- c("title", "source_name_ch1", "characteristics_ch1", "characteristics_ch1.2")
pd <- pd[, keepCols]
colnames(pd) <- c("ptid", "cellType", "infectionStatus", "treatment")

pd$ptid <- sapply(pd$ptid, FUN = function(x) {unlist(strsplit(as.character(x), "_"))[2]}) 
pd$infectionStatus <- gsub(".*: ", "", pd$infectionStatus) 
pd$treatment <- gsub(".*: ", "", pd$treatment)
pd$treatment <- gsub("Poly IC H", "Poly", pd$treatment)
pd
}
```

## Sanityze and select data

We will select the macrophage data as in the paper (Figure 2).
```{r}
gds_new <- gds
pData(gds_new) <- sanitize_pdata(pData(gds_new))
macrophageData <- gds_new[, pData(gds_new)$cellType == "Monocyte-derived Macrophage"]
```

## Look for differentially expressed genes upon stimulation

```{r}
# Load the limma package
suppressMessages(library(limma))

# Set up design matrix
mm1 <- model.matrix(~treatment + ptid, macrophageData)
fit1 <- lmFit(macrophageData, mm1)
ebay1 <- eBayes(fit1)

# Select differentially expressed genes
# According to the paper: Poly(IC) response signatures were generated based on FC >= 1.5 relative to Mock, statistically significant change in exp as determined by limma (BH FDR <0.05)
# BH is default
tt1 <- topTable(ebay1, coef = "treatmentPoly", number = 10,  p.value = 0.01, lfc = log2(1.5), sort.by = "p")
```

## EDA of differentially expressed genes

Let's get our data ready

```{r}
fd <- data.table(fData(gds_new), keep.rownames = TRUE)
setnames(fd, "rn", "probe_name")
pd <- data.table(pData(gds_new), keep.rownames = TRUE)
setnames(pd, "rn", "filename")
ed <- data.table(t(exprs(gds_new)), keep.rownames = TRUE)
setnames(ed, "rn", "filename")
setkey(pd, filename)
setkey(ed, filename)
md <- ed[pd]
```

## Reshaping data

```{r}
library(reshape2)
md_long <- melt(md, variable.name = "probe_name", value.name = "expression_value")
# Add gene variance
md_long[, sd_probe := sd(expression_value), by = probe_name]
```

## Filter and join

```{r}
setkey(md_long, probe_name)
setkey(fd, probe_name)
md_long_short <- fd[md_long[probe_name %in% tt1$Probe_Id] , nomatch = 0]
```


## EDA of DE genes

```{r}
library(ggplot2)
ggplot(md_long_short, aes(x = treatment, y = expression_value)) + geom_violin() + geom_jitter(aes(color = infectionStatus)) + geom_line(aes(group = ptid), alpha = .5) + facet_wrap( ~Symbol + probe_name, scales = "free")
```

**Exercise:** Repeat this with different gene names and geometries


